### ISSUE TITLE
fetch page on back/forward (and keep caching)?

### ISSUE THREAD

**Issue by gingerlime** [turbolinks/turbolinks#413]
Turbolinks looks awesome, and the caching implementation also seems so clever. Also the codebase is so nice and clean. It's a pleasure to read.

We want to use caching, but when the user hits the back/forward, we hope to still be able to load the page (after the initial preview from cache -- just like normal clicks work). The main reason for this, is that we can fetch fresh data from [gon](https://github.com/gazay/gon), which is embedded on the body.

Is there anything we can do to hook into events and detect/distinguish normal visits from back/forward visits? and if we do detect those, can we trigger a real visit *before* `turbolinks:load` fires? Otherwise, is there a way to make browser back/forward behave the same as normal turbolinks visits?

**Comment by domchristie** [comment]
Interesting problem! For pages that are likely to change I have tended to just set `<meta name="turbolinks-cache-control" content="no-cache">` and take the performance hit, however refreshing on back/forward is intriguing.

For restoration visits (i.e. back/forward), Turbolinks binds to the `popstate` event, so as a possible approach, we could _also_ bind to `popstate` and reload the page with Turbolinks. This might look like:

```js
;(function () {
  var reloadWithTurbolinks = (function () {
    var scrollPosition

    function reload () {
      scrollPosition = [window.scrollX, window.scrollY]
      Turbolinks.visit(window.location.toString(), { action: 'replace' })
    }

    window.addEventListener('turbolinks:load', function () {
      if (scrollPosition) {
        window.scrollTo.apply(window, scrollPosition)
        scrollPosition = null
      }
    })

    return reload
  })()

  window.addEventListener('popstate', function () {
    requestAnimationFrame(reloadWithTurbolinks)
  })
})()
```

A couple of downsides. `popstate` is also fired when visiting same-page anchors e.g. `<a href="#my_anchor">Jump to …</a>`, which might be problematic. Secondly there seems to be [a bug with the scroll position when navigating back](https://github.com/turbolinks/turbolinks/issues/333), which is not ideal … but this is a start and it may work for you!

**Comment by gingerlime** [comment]
Thanks for the detailed response and suggestion. 

We also considered hooking into the `popstate`, but we also have some other cases that hook into it (specifically, with Algolia search), and we weren't entirely sure about the best action to take. Looks like `Turbolink.visit(..., {action: 'replace'})` would make this seamless, similarly to what happens with normal page visits. Looks like the anchor link behaviour might throw another spanner in the works in this case :-/

Re adding `no-cache` - in our case, `gon` acts as some kind of bridge between server and client, so it can get updated on almost any page.

I'm curious if the different behaviour on back/forward compared to normal page visit is intentional, or due to how the browser behaves. Do you happen to know? (i.e. normal visit would preview from cache, and then fetch the page and update it. Back/Forward only shows the cached data).

/cc @joker-777

**Comment by domchristie** [comment]
> We also considered hooking into the popstate, but we also have some other cases that hook into it (specifically, with Algolia search), and we weren't entirely sure about the best action to take. Looks like Turbolink.visit(..., {action: 'replace'}) would make this seamless, similarly to what happens with normal page visits, which is more or less what we want.

Yes, this method is pretty much what happens internal in Turbolinks when performing an Application Visit to a page already in the cache: 1. load page from cache 2. visit same location and replace the result. Step 1 is already performed by Turbolinks on `popstate`

> Looks like the anchor link behaviour might throw another spanner in the works in this case unfortunately :-/

Yes this could get a bit tricky. Turbolinks currently makes a request to same-page anchors, so you might get away with it, although this is a bug (dealt with in #285), so its behaviour might change.

> Re adding no-cache - in our case, gon acts as some kind of bridge between server and client, so it can get updated on almost any page.

I suppose if you're working on an app with lots of changes in realtime, by adding `no-cache` in lots of places, you'll be losing a key benefit of Turbolinks. FWIW I work on a messaging app. We use `no-cache` on the list of "chats", so that if someone is viewing a single chat then navigates Back, the list is updated with the latest excerpt (if that makes sense!).

> I'm curious if the different behaviour on back/forward compared to normal page visit is intentional, or due to how the browser behaves. Do you happen to know

I think this is the intended behaviour and maybe inspired by the default browser behaviour:

> Turbolinks automatically initiates a restoration visit when you navigate with the browser’s Back or Forward buttons. Applications using the iOS or Android adapters initiate a restoration visit when moving backward in the navigation stack.
>
> If possible, Turbolinks will render a copy of the page from cache without making a request. Otherwise, it will retrieve a fresh copy of the page over the network. See Understanding Caching for more details.
> https://github.com/turbolinks/turbolinks#restoration-visits

**Comment by gingerlime** [comment]
Thanks again @domchristie! I really appreciate it. We'll need to see what the best approach might be. The documentation is very clear, and matches what we saw in the browser. What I wasn't so sure about was  the *why*, or whether this was the intention. Also, in case we missed something regarding the key events within the turbolinks lifecycle. I can definitely see why it makes sense to avoid another request if you're "just" browsing back into the history.

In our case it might introduce some issues, so we were hoping for some degree of control. Looks like using `popstate` and `.visit(..., {action: 'replace'})` are the best options at the moment.

**Comment by joker-777** [comment]
Thanks, @domchristie for your suggestion. This could definitely be a solution. Best would be though to have the possibility to define through a configuration if I want to have a "restore" visit when clicking on the back button or a normal visit.

**Comment by joker-777** [comment]
Would you guys be interested if we would create a pull request with a possible solution to configure this? /c @domchristie

**Comment by domchristie** [comment]
Feel free to explore this :) I'm not 100% sure this is something we'll accept because `no-cache` is satisfactory in many cases, and although it's not official, or quite the same as an Application visit, the suggested workaround might do the trick. I'll keep this issue open as a place to discuss this, and for others to share their views. Thanks!

**Comment by gingerlime** [comment]
I understand this may seem like some kind of an edge case, but to some extent I think it's even a saner default. I'll try to give some reasoning.

It can eliminate the need for `no-cache` almost entirely if you know that visiting a cache page would immediately load a fresh version. So you get all the benefits of perceived performance, and still get fresh data quickly enough ... Also it increases consistency in behaviour. Instead of caching behaving differently with back/forward and with clicks on cached pages, all pages behave the same.

In any case, we'll look at using the workaround, but we just thought that maybe it's an improvement that others can enjoy (at least as some kind of a configurable option). I'm also very curious what others think about this.
